### DISCUSSION TITLE
why x11 client copy area in EndPaint?

### DISCUSSION THREAD

**Discussion by RyanCode** [FreeRDP/FreeRDP#8968] [Q&A] [answered]
asking a few questions that puzzled me, why did freerdp do something similar to copyImg in the endpaint in the client implementation of x11 which is already implemented a series of expanded drawing functions such as gdi ? And when will endpaint be called? If software gdi is set, will the endpaint callback be the only way to capture bitmap formatted window graphics?Looking forward to your reply.

**Comment by akallabeth** [comment]
the `EndPaint` callback in the client implementation is exactly that, painting has been completed.
the painting is usually done to some backbuffer and the copy operation puts it on screen.

the other operations you might have mistaken for the same are internal ones that draw parts to the internal backbuffer (and might be overwritten if a client provides optimized implementations)

**Reply by RyanCode** [reply]
What you mean that those draw operations in xf_ gdi or xf_graphic are all drawn to the internal cache?Where is EndPaint actively called?Why not directly draw to XServer?I have found that many other RDP client drawings are done in this way, would it be more efficient?

**Reply by akallabeth** [reply, marked as answer]
no. bitmaps are bitmaps and you need to decode these first.
As mentioned before, try to understand the basic operation before diving into optimization.
